Fix/2593 disabling woopay express avoids disabling direct checkout

lovo-h opened the pull request:
Fixes 2593-gh-Automattic/woopay.

#### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

This PR ensures that the WooPay Direct Checkout flow continues to work as intended whether the WooPay Express Checkout button is enabled or disabled on the cart page.

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

**Regression Test: Ensure WooPay Direct Checkout flow works as intended when the Express Checkout button is enabled on the cart page**
* Ensure the WooPay Direct Checkout flow is enabled on the merchant's site.
* As a merchant, navigate to `Payments > Settings`.
* Scroll down to the `Express checkouts` section and click on the `Customize` button for WooPay.
* Under the `Enable WooPay button on selected pages` section, make sure the `Cart Page` is enabled.
* As a shopper, ensure you are authenticated on WooPay.
* Navigate to the Merchant shop, add an item to your cart, and navigate to the Cart page.
* Click the `Proceed to checkout` button.
* Ensure you are navigated to WooPay checkout.
* Click the `Place order` button and ensure you are able to checkout successfully.
* Repeat these steps once using the wcblocks-cart and once using the shortcode-cart.

**Test: Ensure WooPay Direct Checkout flow works as intended when the Express Checkout button is disabled on the cart page**
* Ensure the WooPay Direct Checkout flow is enabled on the merchant's site.
* As a merchant, navigate to `Payments > Settings`.
* Scroll down to the `Express checkouts` section and click on the `Customize` button for WooPay.
* Under the `Enable WooPay button on selected pages` section, make sure the `Cart Page` is disabled.
* As a shopper, ensure you are authenticated on WooPay.
* Navigate to the Merchant shop, add an item to your cart, and navigate to the Cart page.
* Click the `Proceed to checkout` button.
* Ensure you are navigated to WooPay checkout.
* Click the `Place order` button and ensure you are able to checkout successfully.
* Repeat these steps once using the wcblocks-cart and once using the shortcode-cart.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

Commit:
'maybe_enable_woopay_direct_checkout' to 'maybe_init_woopay_direct_checkout'

Before, the function name implied it might enable the Direct Checkout
flow. Which was not accurate since enablement is controlled elsewhere.
Now, the function name implies it might initialize the Direct Checkout
flow. Initialization of the flow is dependent on whether it's enabled.

Commit:
Make Direct Checkout flow independent from First Party Auth flow

Commit:
Code cleanup: Remove unused class member 'express_checkout_helper'

Commit:
Init wcpayConfig when Express Checkout disabled & Direct Checkout enabled

This applies only in the cart page.

Before, when the WooPay Express Checkout button was disabled from the
cart page it would also unintentionally disable the WooPay Direct
Checkout flow. Now, when the WooPay Express Checkout button is disabled
from the cart page, the WooPay Direct Checkout flow continues to
function.

Commit:
Make express_checkout_helper injectable in WC_Payments

Commit:
Code cleanup: create 'enqueue_woopay_common_config_script()' function

Commit:
Add tests to ensure wcpayConfig is loaded only once

wcpayConfig is loaded when:
- On the cart page, when the WooPay Express button is available in the
  cart page.
- On the cart page, when the WooPay Express button is not available in
  the cart page but WooPay Direct Checkout is enabled.

Commit:
Add changelog

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `8440` or branch name `fix/2593-disabling-woopay-express-avoids-disabling-direct-checkout` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=fix/2593-disabling-woopay-express-avoids-disabling-direct-checkout) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: fb1a85d3d3f4ed62946afb5bb646e130d2bed942
- Build time: 2024-03-22 15:20:34 UTC

_Note: the build is updated when a new commit is pushed to this PR._

Commit:
Fix: Button separator doesn't appear when Express Checkout button doesn't

Commit:
Ensure Direct Checkout flow accounts for adapted extensions

Since the First Party Auth already accounts for adapted extensions,
using the same enabling mechanism suffices for now.

bborman22 committed:
Merge branch 'develop' into fix/2593-disabling-woopay-express-avoids-disabling-direct-checkout

bborman22 reviewed (approved):
This tested really well for me after our discussion. To recap that discussion there were two points:
- The OR separator was still being displayed on the cart block. To address this we added the `shouldShowWooPayButton` flag to determine whether to register the express payment method. By not registering it, we avoid the separator being displayed.
- Disabling the direct checkout feature if there are adapted extensions. We got this ability back by keeping the `self:: is_woopay_first_party_auth_enabled` check in the direct checkout feature check. You correctly noted, we'd like to decouple this in the future.

I did discover one additional thing that isn't a total blocker, but within the common config script, could we add the `woopayMinimumSessionData`? While the front-end has a fallback to request it from the merchant server, there is a performance boost by having it available on page load. Not a deal breaker if we might not have access to `WooPay_Session` conveniently there, but wanted to mention it.

Otherwise everything looked good, I tested the following matrix of flows:
- Button displayed on cart page or not
- Cart block and cart shortcode
- Logged into WooPay or not
- Third party cookies blocked or not
- Doing a regular checkout with CC
- Doing a regular checkout with Afterpay

And found no functional issues. 

I'll also note our discussion that `wcpayConfig` setting is very scattered and this would be a prime area to refactor to consolidate all of our settings into one location/variable and allow different features to add/remove/update as needed.
